lib/mongoid/fields.rb

```diff
@@ -310,17 +348,52 @@ def database_field_name(name)
           return nil unless name.present?
           key = name.to_s
           segment, remaining = key.split('.', 2)
-          segment = aliased_fields[segment]&.dup || segment
+
+          # If it's a reference association and there's still remaining, don't
+          # get the alias for this. This is going to retrieve the _id field for
+          # the association name, and getting another property from the _id field
+          # is not the intended functionality.
+          unless remaining && relations.key?(segment) && !relations[segment].embedded?
+            segment = aliased_fields[segment]&.dup || segment
+          end
```

Comment:
Does this logic make sense here? It fixes my tests but I'm not sure it's the right thing. See the comment above it.

Reply:
I think this would be more readable with `if` instead of `unless`.